### PR TITLE
Create a TypeScript program

### LINKED ISSUE #106: Create a TypeScript program
After implementing #101, we'll have three mock programs in the `test/` directory. We should add a fourth program, called `typescript`, with the following structure:

```
test/
   programs/
      typescript/
         cli/
            entry.ts
            entry.json
         package.json
         ...any other files necessary for compiling TypeScript...
```

The `entry.json` file just needs to be empty: `{}`. It doesn't need any special options or flags.

We'll need to add TypeScript related files to this program as if it were its own, completely separate program built on Waterfall CLI and written in TypeScript. That means it will need its own tsconfig.json file, it will actually need appropriate dependencies in the `package.json` file, etc.

Then, we'll need to update the `npm run test` command to ensure that it builds the `typescript` program independently.

Finally, we need to add a `test/typescript.test.js` file that resembles the `es6-imports.test.js` file, except that it doesn't use any experimental Node.js flags. Its entry file should be the compiled version of `entry.ts`.

### PR DESCRIPTION
**Pull request by pbelbin**
Closes #106 

This relaxes some of the recommended TypeScript compilation rules, since the target module (waterfall-cli) is currently *not* a TypeScript source module, so, has no available type declarations.

Note:

This currently creates a JavaScript source from a TypeScript source, and the output (.js and .map.js) files are put into the same folder as the source (.ts).

The main down-side to this is that it makes it more tricky to easily do a 'clean' type operation, as now, the output is co-mingled with the source.

Something a bit better would be good - where the needed files could be copied to another folder along with the compiled output.

### CODE REVIEW

File: test/programs/typescript/tsconfig.json [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+		"module": "commonjs",
+		"moduleResolution": "node",
+		"noStrictGenericChecks": true,
+		"noUnusedLocals": true,
+		"outDir": "cli",
+		"removeComments": true,
+		"resolveJsonModule": true,
+		"rootDir": "cli",
+		"sourceMap": true,
+		//"strict": true,
```
Comment (WesCossick) [inline test/programs/typescript/tsconfig.json]: JSON does not support comments.

**Comment by WesCossick** [comment]
What do you mean when you say, "This relaxes some of the recommended TypeScript compilation rules?"

**Comment by WesCossick** [comment]
> it more tricky to easily do a 'clean' type operation

What do you mean by this?

**Comment by WesCossick** [comment]
> Something a bit better would be good - where the needed files could be copied to another folder along with the compiled output.

What if the `.ts` source files were contained in a folder called `src/` instead of `cli/`, and the compiled files got placed in a folder called `cli/`?

**Comment by pbelbin** [comment]
> > makes it more tricky to easily do a 'clean' type operation
> 
> What do you mean by this?

In the typical case, the transpiled output goes into a folder like `./dist`, so you can easily remove the entire folder without being too concerned.

Currently, in the case of the test typescript program, the transpiled output is relying on being in the same folder as the .json file for it to function correctly.  And, the .json file is not touched by the transpiler.  So, there isn't an output folder that contains entirely items that can simply be removed by removing the entire folder.

I suppose this could be remedied by adding a copy command after the `tsc` command, to copy the .json from src to the folder where the transpiler puts the output.

**Comment by pbelbin** [comment]
> > This relaxes some of the recommended TypeScript compilation rules
> 
> What do you mean by this? This PR doesn't appear to relax any existing rules.

When I added the typescript test, I started out with a copy of the project's main tsconfig.json, with only small modifications to do with paths.  This resulted in failure to compile, due to complaints about missing/conflicting type declarations.  I ended up relaxing some of the requirements for the typescript test so that it would complete the transpilation.

**Comment by pbelbin** [comment]
> > Something a bit better would be good - where the needed files could be copied to another folder along with the compiled output.
> 
> What if the `.ts` source files were contained in a folder called `src/` instead of `cli/`, and the compiled files got placed in the folder called `cli/`? So the structure would look like:
> 
> ```
> test/
>    programs/
>       typescript/
>          cli/
>             ...compiled files...
>          src/
>             entry.ts
>             entry.json
>          package.json
>          tsconfig.json
> ```

Regardless of how the source files are organized, you still have to do extra steps, to put a copy of the .json file into the same folder as the transpilation output.

**Comment by pbelbin** [comment]
> > > makes it more tricky to easily do a 'clean' type operation
> > 
> > 
> > What do you mean by this?
> 
> In the typical case, the transpiled output goes into a folder like `./dist`, so you can easily remove the entire folder without being too concerned.
> 
> Currently, in the case of the test typescript program, the transpiled output is relying on being in the same folder as the .json file for it to function correctly. And, the .json file is not touched by the transpiler. So, there isn't an output folder that contains entirely items that can simply be removed by removing the entire folder.
> 
> I suppose this could be remedied by adding a copy command after the `tsc` command, to copy the .json from src to the folder where the transpiler puts the output.

I found out how to get the .json copied over too.  Source is now in ./src and output goes to ./cli (which can be ignored by git)

**Comment by WesCossick** [comment]
> I ended up relaxing some of the requirements for the typescript test so that it would complete the transpilation.

Okay, I understand now. That's fine; it's just a mock project. The important part is it's written in TypeScript, not that it adheres to strict settings.

> I found out how to get the .json copied over too. Source is now in ./src and output goes to ./cli (which can be ignored by git)

Great! 👍🏼

**Pull request merged**